[Feature] - Codecs derivation based on an existing one

gvolpe opened the pull request:
closes #55 

Introducing utility to derive codecs from an existing one given an isomorphism between two types A and B.

kubukoz commented on core/src/test/scala/com/github/gvolpe/fs2redis/codecs/IsoTests.scala:
```diff
@@ -0,0 +1,72 @@
+    }
+
+  // (iso.from compose iso.to) <-> identity[A]
+  def isomorphism[A, B](implicit iso: Iso[A, B]): A => A =
```
Is it still a proper isomorphism if we don't check the reverse roundtrip? (From andthen to)

(That would probably make the String,Long instance unlawful, as it wouldn't roundtrip for non-numeric string input)

kubukoz replied:
Exactly what I was thinking :P

gvolpe commented:
@kubukoz are you familiar with `discipline` and the whole `cats-laws`? I introduced a `SplitEpi` morphism and a bunch of laws but I'm stuck at this error:

```
[error] /workspace/oss/fs2-redis/core/src/test/scala/com/github/gvolpe/fs2redis/codecs/SplitEpiTests.scala:33:27: No implicit view available from cats.laws.IsEq[B] => org.scalacheck.Prop.
[error]       "identity" -> forAll(laws.identity _),
[error]                           ^
[error] /workspace/oss/fs2-redis/core/src/test/scala/com/github/gvolpe/fs2redis/codecs/SplitEpiTests.scala:34:30: No implicit view available from cats.laws.IsEq[A] => org.scalacheck.Prop.
[error]       "idempotence" -> forAll(laws.idempotence _)
[error]                              ^
[error] two errors found
```

I found some explanation about it [here](https://github.com/oscarrenalias/scalacheck-cookbook/blob/master/markdown/common-problem-scenarios.md) but I still have no clue in how to solve it.

Any ideas?

gvolpe commented:
If this works out (once we check the laws are passing) I'm going to ask Rob if he's okay in making use of the stuff he presented here. I added a link to his talk already since AFAIK the code and slides haven't been published yet.

kubukoz commented:
I'm not *that* familiar but I'll give it a look...

On Thu, Jan 3, 2019, 04:20 Gabriel Volpe <notifications@github.com> wrote:

> If this works out (once we check the laws are passing) I'm going to ask
> Rob if he's okay in making use of the stuff he presented here. I added a
> link to his talk already since AFAIK the code and slides haven't been
> published yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gvolpe/fs2-redis/pull/57#issuecomment-451047572>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA2npJCm6cxtDVQTQGvI5JPGe7m2sW7Oks5u_XcFgaJpZM4ZmJBW>
> .
>

kubukoz commented:
There was an implicit conversion from `IsEq[A]` to `Prop` missing. Now it compiles (although some of the tests don't pass yet)

gvolpe commented:
Cool I remember seeing that conversion but it didn't occur to me to import that package, thanks for looking into it!

I fixed the tests and realized that some of the conversions I defined as epimorphisms only make sense as monomorphisms so I added some more tests.

gvolpe commented on core/src/test/scala/com/github/gvolpe/fs2redis/codecs/SplitMorphismTest.scala:
```diff
@@ -0,0 +1,58 @@
+  val stringDoubleEpi: SplitEpi[String, Double] =
+    SplitEpi(s => Try(s.toDouble).getOrElse(0), _.toString)
+
+  val stringLongEpi: SplitEpi[String, Long] =
```
Thinking that maybe these common instances should be provided on `core` so the users can make use of it.

gvolpe commented:
Got the :+1: from Rob to make use of his idea. If no objections I'll merge this @kubukoz .
